File: python/Benedikt/day_2/day_2.py
file = 'input'


def read(in_file):
    dic = {}
    with open(in_file, 'r') as lines:
        for line in lines:
            comp_line = line.strip()
            comp_lst = comp_line.split()
            if comp_lst[0] not in dic and comp_lst[0] != 'up':
                dic[comp_lst[0]] = int(comp_lst[1])
            elif comp_lst[0] == 'up':
                dic['down'] -= int(comp_lst[1])
            else:
                dic[comp_lst[0]] += int(comp_lst[1])
    # print(dic)
    horizontal = dic['forward']
    depth = dic['down']
    return horizontal * depth


def main(file):
    result = read(file)
    print("The final result of the multiplied horizontal and depth\
numbers is: ", result)


if __name__ == '__main__':
    main(file)

File: haskell/jherrlin/day2/day2.hs
data Direction = Up Int | Down Int | Forward Int deriving (Show)

readLine :: [String] -> (String, Int)
readLine [a, b] = (a, read b)

toDirection :: (String, Int) -> Direction
toDirection ("forward", x) = Forward x
toDirection ("down", x)    = Down x
toDirection ("up", x)      = Up (-x)

horizontalPred :: Direction -> Bool
horizontalPred (Forward _) = True
horizontalPred _ = False

verticalPred :: Direction -> Bool
verticalPred (Up _) = True
verticalPred (Down _) = True
verticalPred _ = False

directionToInt :: Direction -> Int
directionToInt (Up x) = x
directionToInt (Down x) = x
directionToInt (Forward x) = x

problem1 = do
  contents <- readFile "input.txt"
  let directions  = map (toDirection . readLine . words) $ lines contents
      horizontals = sum $ map directionToInt $ filter horizontalPred directions
      verticals   = sum $ map directionToInt $ filter verticalPred directions
      result      = horizontals * verticals
  print result


acc :: (Int, Int, Int) -> (String, Int) -> (Int, Int, Int)
acc (horizontal, depth, aim) ("down", x)    = (horizontal,     depth,             aim + x)
acc (horizontal, depth, aim) ("up", x)      = (horizontal,     depth,             aim - x)
acc (horizontal, depth, aim) ("forward", x) = (horizontal + x, depth + (aim * x), aim)

problem2 = do
  contents <- readFile "input.txt"
  let directions             = map (readLine . words) $ lines contents
      (horizontal, depth, _) = foldl acc (0, 0, 0) directions
      result                 = horizontal * depth
  print result
